File: android/ac-templating/src/main/kotlin/com/microsoft/adaptivecards/templating/ExpressionEvaluator.kt
package com.microsoft.adaptivecards.templating

import com.microsoft.adaptivecards.templating.functions.*

/**
 * Expression function interface
 */
interface ExpressionFunction {
    fun call(arguments: List<Any?>): Any?
}

/**
 * Evaluates parsed expressions against a data context
 */
class ExpressionEvaluator(private val context: DataContext) {
    private val functions: Map<String, ExpressionFunction>

    init {
        val functionsMap = mutableMapOf<String, ExpressionFunction>()

        // Register all built-in functions
        StringFunctions.register(functionsMap)
        DateFunctions.register(functionsMap)
        CollectionFunctions.register(functionsMap)
        LogicFunctions.register(functionsMap)
        MathFunctions.register(functionsMap)

        functions = functionsMap
    }

    /**
     * Evaluate an expression
     * @param expression The parsed expression
     * @return The evaluated result
     * @throws EvaluationException if evaluation fails
     */
    fun evaluate(expression: Expression): Any? {
        return when (expression) {
            is Expression.Literal -> expression.value

            is Expression.PropertyAccess -> context.resolve(expression.path)

            is Expression.FunctionCall -> {
                val function = functions[expression.name]
                    ?: throw EvaluationException("Unknown function: ${expression.name}")

                val evaluatedArgs = expression.arguments.map { evaluate(it) }
                try {
                    function.call(evaluatedArgs)
                } catch (e: Exception) {
                    throw EvaluationException("Function ${expression.name} failed: ${e.message}", e)
                }
            }

            is Expression.BinaryOp -> evaluateBinaryOp(expression.operator, expression.left, expression.right)

            is Expression.UnaryOp -> evaluateUnaryOp(expression.operator, expression.operand)

            is Expression.Ternary -> {
                val condResult = evaluate(expression.condition)
                val boolResult = coerceToBool(condResult)
                evaluate(if (boolResult) expression.trueValue else expression.falseValue)
            }
        }
    }

    // MARK: - Binary Operations

    private fun evaluateBinaryOp(op: String, left: Expression, right: Expression): Any? {
        val leftValue = evaluate(left)
        val rightValue = evaluate(right)

        return when (op) {
            "+" -> {
                // String concatenation or numeric addition
                if (leftValue is String || rightValue is String) {
                    stringValue(leftValue) + stringValue(rightValue)
                } else {
                    coerceToNumber(leftValue) + coerceToNumber(rightValue)
                }
            }
            "-" -> coerceToNumber(leftValue) - coerceToNumber(rightValue)
            "*" -> coerceToNumber(leftValue) * coerceToNumber(rightValue)
            "/" -> {
                val divisor = coerceToNumber(rightValue)
                if (divisor == 0.0) {
                    throw EvaluationException("Division by zero")
                }
                coerceToNumber(leftValue) / divisor
            }
            "%" -> {
                val divisor = coerceToNumber(rightValue)
                if (divisor == 0.0) {
                    throw EvaluationException("Division by zero")
                }
                coerceToNumber(leftValue) % divisor
            }
            "==" -> isEqual(leftValue, rightValue)
            "!=" -> !isEqual(leftValue, rightValue)
            "<" -> coerceToNumber(leftValue) < coerceToNumber(rightValue)
            ">" -> coerceToNumber(leftValue) > coerceToNumber(rightValue)
            "<=" -> coerceToNumber(leftValue) <= coerceToNumber(rightValue)
            ">=" -> coerceToNumber(leftValue) >= coerceToNumber(rightValue)
            "&&" -> coerceToBool(leftValue) && coerceToBool(rightValue)
            "||" -> coerceToBool(leftValue) || coerceToBool(rightValue)
            else -> throw EvaluationException("Unknown operator: $op")
        }
    }

    private fun evaluateUnaryOp(op: String, operand: Expression): Any? {
        val value = evaluate(operand)

        return when (op) {
            "!" -> !coerceToBool(value)
            "-" -> -coerceToNumber(value)
            else -> throw EvaluationException("Unknown operator: $op")
        }
    }

    // MARK: - Type Coercion

    private fun coerceToNumber(value: Any?): Double {
        return when (value) {
            is Double -> value
            is Int -> value.toDouble()
            is Long -> value.toDouble()
            is Float -> value.toDouble()
            is String -> value.toDoubleOrNull() ?: 0.0
            is Boolean -> if (value) 1.0 else 0.0
            null -> 0.0
            else -> throw EvaluationException("Cannot convert $value to number")
        }
    }

    private fun coerceToBool(value: Any?): Boolean {
        return when (value) {
            is Boolean -> value
            is Double -> value != 0.0
            is Int -> value != 0
            is Long -> value != 0L
            is Float -> value != 0.0f
            is String -> value.isNotEmpty()
            null -> false
            is List<*> -> value.isNotEmpty()
            is Map<*, *> -> value.isNotEmpty()
            else -> true // Non-null objects are truthy
        }
    }

    private fun isEqual(left: Any?, right: Any?): Boolean {
        // Handle nil cases
        if (left == null && right == null) return true
        if (left == null || right == null) return false

        // Try numeric comparison
        if (left is Number && right is Number) {
            return left.toDouble() == right.toDouble()
        }

        // Try string comparison
        if (left is String && right is String) {
            return left == right
        }

        // Try boolean comparison
        if (left is Boolean && right is Boolean) {
            return left == right
        }

        // Fallback to string representation
        return left.toString() == right.toString()
    }

    private fun stringValue(value: Any?): String {
        return when (value) {
            is String -> value
            is Double -> {
                // Format numbers without unnecessary decimal places
                if (value % 1.0 == 0.0) {
                    value.toInt().toString()
                } else {
                    value.toString()
                }
            }
            is Int -> value.toString()
            is Boolean -> value.toString()
            null -> ""
            else -> value.toString()
        }
    }
}

/**
 * Exception thrown when expression evaluation fails
 */
class EvaluationException(message: String, cause: Throwable? = null) : Exception(message, cause)

File: ios/Sources/ACTemplating/ExpressionEvaluator.swift
import Foundation

/// Evaluates parsed expressions against a data context
public final class ExpressionEvaluator {
    private let context: DataContext
    private let functions: [String: ExpressionFunction]

    /// Shared function registry built once and reused across all evaluator instances
    private static let sharedFunctions: [String: ExpressionFunction] = {
        var functions: [String: ExpressionFunction] = [:]

        // String functions
        StringFunctions.register(into: &functions)

        // Date functions
        DateFunctions.register(into: &functions)

        // Collection functions
        CollectionFunctions.register(into: &functions)

        // Logic functions
        LogicFunctions.register(into: &functions)

        // Math functions
        MathFunctions.register(into: &functions)

        return functions
    }()

    public init(context: DataContext) {
        self.context = context
        self.functions = Self.sharedFunctions
    }
    
    /// Evaluate an expression
    /// - Parameter expression: The parsed expression
    /// - Returns: The evaluated result
    /// - Throws: EvaluationError if evaluation fails
    public func evaluate(_ expression: Expression) throws -> Any? {
        switch expression {
        case .literal(let value):
            return value
            
        case .propertyAccess(let path):
            return context.resolve(path: path)
            
        case .functionCall(let name, let arguments):
            guard let function = functions[name] else {
                throw EvaluationError.unknownFunction(name)
            }
            
            let evaluatedArgs = try arguments.map { try evaluate($0) }
            return try function.call(evaluatedArgs)
            
        case .binaryOp(let op, let left, let right):
            return try evaluateBinaryOp(op: op, left: left, right: right)
            
        case .unaryOp(let op, let operand):
            return try evaluateUnaryOp(op: op, operand: operand)
            
        case .ternary(let condition, let trueValue, let falseValue):
            let condResult = try evaluate(condition)
            let boolResult = try coerceToBool(condResult)
            return try evaluate(boolResult ? trueValue : falseValue)
        }
    }
    
    // MARK: - Binary Operations
    
    private func evaluateBinaryOp(op: String, left: Expression, right: Expression) throws -> Any? {
        let leftValue = try evaluate(left)
        let rightValue = try evaluate(right)
        
        switch op {
        case "+":
            // String concatenation or numeric addition
            if let l = leftValue as? String, let r = rightValue as? String {
                return l + r
            } else if let l = leftValue as? String {
                return l + String(describing: rightValue ?? "")
            } else if let r = rightValue as? String {
                return String(describing: leftValue ?? "") + r
            }
            let left = try coerceToNumber(leftValue)
            let right = try coerceToNumber(rightValue)
            return left + right

        case "-":
            let left = try coerceToNumber(leftValue)
            let right = try coerceToNumber(rightValue)
            return left - right

        case "*":
            let left = try coerceToNumber(leftValue)
            let right = try coerceToNumber(rightValue)
            return left * right
            
        case "/":
            let divisor = try coerceToNumber(rightValue)
            guard divisor != 0 else {
                throw EvaluationError.divisionByZero
            }
            return try coerceToNumber(leftValue) / divisor
            
        case "%":
            let divisor = try coerceToNumber(rightValue)
            guard divisor != 0 else {
                throw EvaluationError.divisionByZero
            }
            return try coerceToNumber(leftValue).truncatingRemainder(dividingBy: divisor)
            
        case "==":
            return isEqual(leftValue, rightValue)
            
        case "!=":
            return !isEqual(leftValue, rightValue)
            
        case "<":
            let left = try coerceToNumber(leftValue)
            let right = try coerceToNumber(rightValue)
            return left < right

        case ">":
            let left = try coerceToNumber(leftValue)
            let right = try coerceToNumber(rightValue)
            return left > right

        case "<=":
            let left = try coerceToNumber(leftValue)
            let right = try coerceToNumber(rightValue)
            return left <= right

        case ">=":
            let left = try coerceToNumber(leftValue)
            let right = try coerceToNumber(rightValue)
            return left >= right

        case "&&":
            let left = try coerceToBool(leftValue)
            let right = try coerceToBool(rightValue)
            return left && right

        case "||":
            let left = try coerceToBool(leftValue)
            let right = try coerceToBool(rightValue)
            return left || right
            
        default:
            throw EvaluationError.unknownOperator(op)
        }
    }
    
    private func evaluateUnaryOp(op: String, operand: Expression) throws -> Any? {
        let value = try evaluate(operand)
        
        switch op {
        case "!":
            return try !coerceToBool(value)
            
        case "-":
            return try -coerceToNumber(value)
            
        default:
            throw EvaluationError.unknownOperator(op)
        }
    }
    
    // MARK: - Type Coercion
    
    private func coerceToNumber(_ value: Any?) throws -> Double {
        if let num = value as? Double {
            return num
        } else if let num = value as? Int {
            return Double(num)
        } else if let num = value as? Float {
            return Double(num)
        } else if let str = value as? String, let num = Double(str) {
            return num
        } else if let bool = value as? Bool {
            return bool ? 1.0 : 0.0
        } else if value == nil {
            return 0.0
        }
        
        throw EvaluationError.typeCoercionFailed(value, "number")
    }
    
    private func coerceToBool(_ value: Any?) throws -> Bool {
        if let bool = value as? Bool {
            return bool
        } else if let num = value as? Double {
            return num != 0
        } else if let num = value as? Int {
            return num != 0
        } else if let str = value as? String {
            return !str.isEmpty
        } else if value == nil {
            return false
        } else if let array = value as? [Any] {
            return !array.isEmpty
        } else if let dict = value as? [String: Any] {
            return !dict.isEmpty
        }
        
        return true // Non-null objects are truthy
    }
    
    private func isEqual(_ left: Any?, _ right: Any?) -> Bool {
        // Handle nil cases
        if left == nil && right == nil {
            return true
        }
        if left == nil || right == nil {
            return false
        }
        
        // Try numeric comparison
        if let l = left as? Double, let r = right as? Double {
            return l == r
        }
        if let l = left as? Int, let r = right as? Int {
            return l == r
        }
        
        // Try string comparison
        if let l = left as? String, let r = right as? String {
            return l == r
        }
        
        // Try boolean comparison
        if let l = left as? Bool, let r = right as? Bool {
            return l == r
        }
        
        // Fallback to string representation
        return String(describing: left ?? "nil") == String(describing: right ?? "nil")
    }
}

// MARK: - Expression Function Protocol

public protocol ExpressionFunction {
    func call(_ arguments: [Any?]) throws -> Any?
}

// MARK: - Errors

public enum EvaluationError: Error, LocalizedError {
    case unknownFunction(String)
    case unknownOperator(String)
    case typeCoercionFailed(Any?, String)
    case divisionByZero
    case invalidArgumentCount(expected: Int, actual: Int)
    case invalidArgument(String)
    
    public var errorDescription: String? {
        switch self {
        case .unknownFunction(let name):
            return "Unknown function: \(name)"
        case .unknownOperator(let op):
            return "Unknown operator: \(op)"
        case .typeCoercionFailed(let value, let type):
            return "Cannot convert \(String(describing: value)) to \(type)"
        case .divisionByZero:
            return "Division by zero"
        case .invalidArgumentCount(let expected, let actual):
            return "Invalid argument count: expected \(expected), got \(actual)"
        case .invalidArgument(let message):
            return "Invalid argument: \(message)"
        }
    }
}
